src/identity_tree.rs

```diff
@@ -28,181 +34,340 @@ impl TreeUpdate {
-            let mut next = next.0.write().await;
-
-            // Get the updates to be applied and apply them sequentially. It is very
-            // important that we record the merkle proof after each step as they depend on
-            // each other.
-            let updates: Vec<&TreeUpdate> = next.diff.iter().take(update_count).collect();
-            for update in &updates {
-                self.update(update.leaf_index, update.element);
-                let proof = self.tree.proof(update.leaf_index).unwrap_or_else(|| {
-                    panic!("No proof exists at leaf index {}.", update.leaf_index);
-                });
-                proofs.push(proof);
+            {
+                // Acquire the exclusive write lock on the next version.
+                let mut next = next.get_data();
+
+                // Get the updates to be applied and apply them sequentially. It is very
+                // important that we record the merkle proof after each step as they depend on
+                // each other.
+                let updates: Vec<&TreeUpdate> =
+                    next.metadata.diff.iter().take(update_count).collect();
+                for update in &updates {
+                    self.update(update.leaf_index, update.element);
+                    let proof = self.tree.proof(update.leaf_index);
+                    proofs.push(proof);
+                }
+
+                // Remove only the updates that have been consumed, which may be all of them.
+                next.metadata.diff = if next.metadata.diff.len() > updates.len() {
+                    Vec::from(&next.metadata.diff[updates.len()..])
+                } else {
+                    vec![]
+                }
             }
+            self.garbage_collect();
+        }
+        proofs
+    }
+}
 
-            // Remove only the updates that have been consumed, which may be all of them.
-            next.diff = if next.diff.len() > updates.len() {
-                Vec::from(&next.diff[updates.len()..])
-            } else {
-                vec![]
+impl BasicTreeOps for TreeVersionData<lazy_merkle_tree::Canonical> {
+    fn update(&mut self, leaf_index: usize, element: Hash) {
+        utils::replace_with(&mut self.tree, |tree| {
+            tree.update_with_mutation(leaf_index, &element)
+        });
+        self.next_leaf = leaf_index + 1;
+        self.metadata.count_since_last_flatten += 1;
+    }
+
+    /// Garbage collection for the canonical tree version. It rewrites all
+    /// future versions of the tree to use the more optimized storage of this
+    /// tree. This is done periodically, to really make the additional
+    /// time-costs of these rebuilds repay themselves in saved memory.
+    ///
+    /// Warning: this will attempt to acquire lock for all transitive successors
+    /// of this tree, and therefore no version locks acquired through
+    /// `TreeVersion#get_data()` may be held at the time of calling this.
```

Comment:
We likely can't do better in the type system, but it might be good to try to lock the lock and then raise an error before backing out or dying.

Reply:
This lock can be congested due to legitimate reasons, this comment is just so that we don't deadlock internally.